taskqueue/gcloud/aio/taskqueue/taskmanager.py

```diff
@@ -2,486 +2,159 @@
-            fire(self.delete, id_)
-            self.completed_tasks += 1
-
-            self.sizes.append(size)
+                await self.worker(payload)
+            except FailFastError as e:
+                log.error('[FailFastError] failed to process task: %s', name)
+                log.exception(e)
 
-            if duration is not None:
-                self.durations.append(duration)
+                autorenew.cancel()
+                with contextlib.suppress(asyncio.CancelledError):
+                    await autorenew
 
-            log.info('Finished {{id: %s, size: %s, duration: %ds}}', object_id,
-                     size, duration)
+                task = self.tasks[name]
+                del self.tasks[name]
 
-        def fail(ex):
-
-            self.failed += 1
-
-            log.error('Worker did not complete task %s for %s', id_,
-                      payload['name'])
-            log.exception(ex)
-
-            log.error('Current retry_count: %d', task['retry_count'])
-
-            if not self.deadletter_upsert:
-                self.let_go_of(id_)
+                await self.tq.delete(task)
+                await self.fail(task, payload, e)
                 return
+            except Exception as e:  # pylint: disable=broad-except
+                log.error('failed to process task: %s', name)
+                log.exception(e)
 
-            fail_fast = isinstance(ex, FailFastError)
+                autorenew.cancel()
+                with contextlib.suppress(asyncio.CancelledError):
+                    await autorenew
 
-            if int(task['retry_count']) < self.retry_limit and not fail_fast:
-                self.let_go_of(id_)
-                return
+                task = self.tasks[name]
+                del self.tasks[name]
 
-            log.info('Moving task %s to dead letter queue', id_)
+                tries = task['status']['attemptDispatchCount']
+                if self.retry_limit is None or tries < self.retry_limit:
+                    await self.tq.cancel(task)
+                    return
 
-            do_insert = fire(self.flunk_to_store, id_, payload, ex)
-
-            @do_insert.add_done_callback
-            def after_insert(_future):  # pylint: disable=unused-variable
-
-                self.let_go_of(id_)
-
-        processed.add_done_callback(done)
+                log.warning('exceeded retry_limit, failing task')
+                await self.tq.delete(name)
+                await self.fail(task, payload, e)
+                return
 
-    def auto_renew(self, id_, payload):
+            log.info('successfully processing task: %s', name)
 
-        self.hold_onto(id_, payload)
+            autorenew.cancel()
+            with contextlib.suppress(asyncio.CancelledError):
+                await autorenew
 
-        half_lease = self.lease_seconds / 2
-        call_later(half_lease, self.renew, id_, half_lease)
+            task = self.tasks[name]
+            del self.tasks[name]
```

Comment:
Any reason you're repeating these calls inside and outside the try-except block here? I might suggest something like this:
```
try:
    # code
except FirstError:
    # code
except SecondError:
    # code
finally:
    autorenew.cancel()
    with contextlib.suppress(asyncio.CancelledError):
        await autorenew

    task = self.tasks[name]
    del self.tasks[name]
```

This will trigger prior to the early returns in the `except` blocks, and will also trigger if no exception if hit.

Reply:
Wouldn't the order be incorrect? I need to cancel the `autorenew` thread and grab the task from it after it has been cancelled in order to use that task object for cancelling/acking/deleting.

```
kevin@atomicbox » python3.6
>>> try:
...  raise ValueError(42)
... except Exception:
...  print('except')
... finally:
...  print('finally')
...
except
finally
```

Reply:
Though I would love to de-dupe this block if we can think of a way to get the ordering right.

Reply:
Ah, gotcha, I missed the fact that you were using that return value. I can't think of a function name for the repeated code block so it's not worth pulling out into a function. Otherwise there are no other straight-forward and readable ways to de-dupe this block. This is fine!